source/material_model/visco_plastic.cc

```diff
@@ -0,0 +1,617 @@
+    {
+      // This function calculates viscosities assuming that all the compositional fields
+      // experience the same strain rate (isostrain).
+
+      // Calculate the square root of the second moment invariant for the deviatoric strain rate tensor.
+      // The first time this function is called (first iteration of first time step)
+      // a specified "reference" strain rate is used as the returned value would
+      // otherwise be zero.
+      const double edot_ii = ( (this->get_timestep_number() == 0 && strain_rate.norm() <= std::numeric_limits<double>::min())
+                               ?
+                               ref_strain_rate
+                               :
+                               std::max(std::sqrt(std::fabs(second_invariant(deviator(strain_rate)))),
+                                        min_strain_rate) );
+
+      // Calculate viscosities for each of the individual compositional phases
+      std::vector<double> composition_viscosities(volume_fractions.size());
+      for (unsigned int j=0; j < volume_fractions.size(); ++j)
+        {
+          // Power law creep equation
+          //    viscosity = 0.5 * A^(-1/n) * edot_ii^((1-n)/n) * d^(-m/n) * exp((E + P*V)/(nRT))
+          // A: prefactor, edot_ii: square root of second invariant of deviatoric strain rate tensor,
+          // d: grain size, m: grain size exponent, E: activation energy, P: pressure,
+          // V; activation volume, n: stress exponent, R: gas constant, T: temperature.
+          // Note: values of A, d, m, E, V and n are distinct for diffusion & dislocation creep
+
+          // Diffusion creep: viscosity is grain size dependent (m!=0) and strain-rate independent (n=1)
+          double viscosity_diffusion = 0.5 * std::pow(prefactors_diffusion[j],-1/stress_exponents_diffusion[j]) *
+                                       std::exp((activation_energies_diffusion[j] + pressure*activation_volumes_diffusion[j])/
+                                                (constants::gas_constant*temperature*stress_exponents_diffusion[j])) *
+                                       std::pow(grain_size, -grain_size_exponents_diffusion[j]);
+
+          // For dislocation creep, viscosity is grain size independent (m=0) and strain-rate dependent (n>1)
+          double viscosity_dislocation = 0.5 * std::pow(prefactors_dislocation[j],-1/stress_exponents_dislocation[j]) *
+                                         std::exp((activation_energies_dislocation[j] + pressure*activation_volumes_dislocation[j])/
+                                                  (constants::gas_constant*temperature*stress_exponents_dislocation[j])) *
+                                         std::pow(edot_ii,((1. - stress_exponents_dislocation[j])/stress_exponents_dislocation[j]));
+
+          // Composite viscosity
+          double viscosity_composite = (viscosity_diffusion * viscosity_dislocation)/(viscosity_diffusion + viscosity_dislocation);
```

Comment:
I have a question about your composite diffusion/dislocation rheology:
In the `diffusion_dislocation` material model, we use the dislocation strain rate here instead of `edot_ii` to compute the dislocation viscosity. (v_{eff}=\sigma_{ii}/(2 edot_{ii}), but v_{dis}=\sigma_{ii}/(2 edot_{ii,dis}).) Because we do not have the dislocation strain rate, we have to do some iterations until we haver determined how much of the strain rate is due to diffusion/dislocation creep (the while loop in the `diffusion_dislocation` material model). 
I am not really sure what people normally do, but is there any reason you have done it like this? I guess the additional plasticity would make iterating out which process accommodates the strain even more complicated. 
I was just thinking, if we have two material models that use composite diffusion/dislocation creep, but the implementations are different, we should point that out somewhere and say why they are different. Also, I guess it would be interesting to see how much of a difference it makes. Have you compared your model (without plasticity) to the `diffusion_dislocation` material model?